semtype.lisp

```diff
@@ -0,0 +1,83 @@
+;; Data structure for representing ULF semantic types
+
+(in-package :ulf-lib)
+
+;; Class representing a ULF semantic type
+(defclass semtype ()
+  ((domain
+     :initarg :domain
+     :accessor domain)
+   (range
+     :initarg :range
+     :initform nil
+     :accessor range)
+   (exponent
+     :initarg :ex
+     :initform 1
+     :accessor ex)
+   (subscript
+     :initarg :subscript
+     :initform nil
+     :accessor subscript)
+   (tense?
+     :initarg :tense?
+     :initform nil
+     :accessor tense?)))
+
+;; Check if a given semantic type is an atomic type.
+;; Atomic semantic types have a symbol domain and a nil range.
+(defun semtype-atom? (s)
+  (and (or (symbolp (domain s)) (integerp (domain s))) (not (range s))))
+
+;; Print a given semantic type
+(defun print-semtype (s)
+  (if (semtype-atom? s)
+    (progn
+      (format t "~a" (domain s))
+      (when (> (ex s) 1) (format t "^~a" (ex s))))
+    (progn
+      (format t "(")
+      (print-semtype (domain s))
+      (format t "=>")
+      (print-semtype (range s))
+      (format t ")")
+      (when (subscript s) (format t  "_~a" (subscript s)))
+      (when (tense? s) (format t "_~a" (tense? s)))
+      (unless (equal (ex s) 1) (format t "^~a" (ex s))))))
+
+;; Split a string of form ({domain}=>{range}) into {domain} and {range}
+(defun split-semtype-str (s)
+  (setf level 0)
```

Comment:
Use 
```
(let ((level 0)
       (i 1))
   [body])
```
Instead of `setf` because `setf` initializes a global variable if used on an undeclared variable.